### PR TITLE
Upgrade/fix/simplify store to load forwarding

### PR DESCRIPTION
**Pull request by bondhugula**
- fix store to load forwarding for a certain set of cases (where
  forwarding shouldn't have happened); use AffineValueMap difference
  based MemRefAccess equality checking; utility logic is also greatly
  simplified

- add missing equality/inequality operators for AffineExpr ==/!= ints

- add == != operators on MemRefAccess

### CODE REVIEW

**Comment by River707** [comment]
@andydavis1 Can you review?

File: lib/Analysis/AffineStructures.cpp [hunk 1]
```diff
@@ -2717,51 +2717,6 @@ void FlatAffineConstraints::projectOut(Value *id) {
   FourierMotzkinEliminate(pos);
 }
 
-bool FlatAffineConstraints::isRangeOneToOne(unsigned start,
```
Comment (andydavis1) [inline lib/Analysis/AffineStructures.cpp]: Was this just not used anywhere?

Comment (bondhugula) [reply lib/Analysis/AffineStructures.cpp]: It just had one user and that's now gone with this update.

File: test/Transforms/memref-dataflow-opt.mlir [hunk 2]
```diff
@@ -247,3 +247,35 @@ func @store_load_store_nested_fwd(%N : index) -> f32 {
...
+func @should_not_fwd(%A: memref<100xf32>, %M : index, %N : index) -> f32 {
+  %cf = constant 0.0 : f32
+  affine.store %cf, %A[%M] : memref<100xf32>
+  // CHECK: affine.load %{{.*}}[%{{.*}}]
+  %v = affine.load %A[%N] : memref<100xf32>
+  return %v : f32
+}
+
+// Can store forward to A[%j, %i], but no forwarding to load on %A[%i, %j]
+func @refs_not_known_to_be_equal(%A : memref<100 x 100 x f32>, %M : index) {
```
Comment (River707) [inline test/Transforms/memref-dataflow-opt.mlir]: Missing a CHECK-LABEL here.

Comment (bondhugula) [reply test/Transforms/memref-dataflow-opt.mlir]: Thanks, fixed.

Comment (bondhugula) [reply test/Transforms/memref-dataflow-opt.mlir]: I can't find any information on the failed check below.

**Pull request closed without merging**